hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/SqlConnector.java

```diff
@@ -247,6 +249,22 @@ List<MappingField> resolveAndValidateFields(
             @Nonnull List<MappingField> userFields
     );
 
+    /**
+     * Returns the required permissions to execute
+     * {@link #resolveAndValidateFields(NodeEngine, SqlExternalResource, List)} method.
+     * <p>
+     * Implementors of {@link SqlConnector} don't need to override this method when {@code resolveAndValidateFields}
+     * doesn't support field resolution or when validation doesn't access the external resource.
+     * <p>
+     * The permissions are usually the same as required permissions to read from the external resource.
+     *
+     * @return list of permissions required to run {@link #resolveAndValidateFields}
+     */
+    @Nonnull
+    default List<Permission> permissionsForResolve(SqlExternalResource resource, NodeEngine nodeEngine) {
```

Comment:
we should implement this method for other connectors that query external resources during resolving, eg. IMap (if the columns are not specified), even if the resolved metadata do not contain any secrets but still those are information inaccessible to given user. Other connector do not currently have such feature (Kafka, Mongo)

Reply:
For Kafka we don't have permissions at all yet, it's on a roadmap so this will be implemented as part of HZ-2813

For Mongodb the permissions don't follow the current design. so if we implement it here, it will likely be wrong There is for reviewing/fixing that - see HZ-3019 and HZ-3020.

If you feel strongly about Mongo exposing the info (field names in this case) then we need to implement HZ-3019/HZ-3020 together with this fix.

Reply:
I get the argument for Kafka and Mongo @kwart WDYT?

Still, it should be possible to implement for IMaps.

Reply:
As far as I can see the IMap never resolves the schema from data, always uses the options (e.g. avro/compact schma). So does it make sense to require e.g. read permission to the IMap in question?

Maybe if there were separate permissions for e.g. compact schema, we could check that, but we don't have that right now.

Reply:
> As far as I can see the IMap never resolves the schema from data, always uses the options (e.g. avro/compact schma). So does it make sense to require e.g. read permission to the IMap in question?

you are right, I confused it with `SqlService.mappingDdl` which, by the way, also does not check permissions. And that one gets random entry from IMap and determines schema based on serialization format of the entry.
We also perform that process when handling "missing mapping" error. We helpfully provide command to create mapping with all the fields in the error message.
This can leak information about existence and structure of and IMap that is inaccessible to the user but this is a different issue from what this PR is solving.

Reply:
https://hazelcast.atlassian.net/browse/HZ-3274